Add OIDC login with Authlib (Google, GitHub, Microsoft) – closes #355

yuriBean opened the pull request:
This PR adds backend-only OpenID Connect (OIDC) login to the Gramps Web API using [Authlib](https://docs.authlib.org). It allows users to log in using Google, GitHub, or Microsoft accounts.


Features

- Backend-only OIDC login
- Providers supported: Google, GitHub, Microsoft
- Secure secret handling via `.env`
- Modular `auth/oidc.py` blueprint
- Updates to `README.md` and `.env.example`
Setup

1. Install requirements:

   pip install Authlib python-dotenv

2. Add your .env:
 
   cp .env.example .env

3. Start server:

   flask run

jonade commented:
Thanks for the work on this. What about support for OIDC providers such as Authentik, Keycloak, Authelia, etc? These can allow logins from their own internal users, _or_ be configured from external sources (e.g. from Google, GitHub, or Microsoft accounts). Ideally we should support both scenarios, as not everyone would want to login from these Google etc accounts.

DavidMStraub commented:
> Thanks for the work on this. What about support for OIDC providers such as Authentik, Keycloak, Authelia, etc? These can allow logins from their own internal users, _or_ be configured from external sources (e.g. from Google, GitHub, or Microsoft accounts). Ideally we should support both scenarios, as not everyone would want to login from these Google etc accounts.

I agree, we should definitely add this before shipping this feature, but it might be out of scope of this PR. This should be an MVP demonstrating integration of OIDC with the current authorization system.

yuriBean commented:
The latest commit contains these key changes:

1. Added OIDC authentication endpoints (`/auth/login/<provider>` and `/auth/callback/<provider>`)
2. Integrated with existing JWT-based authentication system
3. Added support for Google, GitHub, and Microsoft as identity providers
4. New users are automatically created with owner role when they first authenticate
5. Tokens include necessary claims (permissions and tree ID) for API access

DavidMStraub commented on requirements-dev.txt:
```diff
@@ -8,3 +8,4 @@ pre-commit
 celery[pytest]
 moto[s3]<5.0.0
 PyYAML
+authlib
```
This is not a dev requirement! It should jo in `pyproject.toml`

DavidMStraub commented on .env.example:
```diff
@@ -0,0 +1,15 @@
+# Google
```
So this file is not used anymore?

DavidMStraub replied:
Then why is it still here?

DavidMStraub commented on gramps_webapi/app.py:
```diff
@@ -1,210 +1,181 @@
-# You should have received a copy of the GNU Affero General Public License
-# along with this program. If not, see <https://www.gnu.org/licenses/>.
-#
-
```
You cannot remove the copyright notice!! (I know ChatGPT likes to do that...)

DavidMStraub commented on gramps_webapi/app.py:
```diff
@@ -1,210 +1,181 @@
-                " variable is deprecated and will stop working in the future."
-                f" Please use `GRAMPSWEB_{option}` instead."
+                f"Setting `{option}` via the environment is deprecated. "
+                f"Use `GRAMPSWEB_{option}` instead."
```
Why did you change this??

DavidMStraub commented on gramps_webapi/app.py:
```diff
@@ -1,210 +1,181 @@
     app.logger.setLevel(logging.INFO)
 
-    # load default config
+
```
If your LLM removes the comments, you need to put them back in

DavidMStraub commented on gramps_webapi/auth/oidc.py:
```diff
@@ -0,0 +1,155 @@
+@oidc_bp.route("/login/<provider>")
+def login(provider):
+    if not current_app.config.get("OAUTH_ENABLED", False):
+        return jsonify({"error": "OAuth is not enabled"}), 403
```
Please use the utility method `abort_with_message`

DavidMStraub commented on gramps_webapi/auth/oidc.py:
```diff
@@ -0,0 +1,155 @@
+
+    email = user_info.get("email")
+    if not email:
+        return jsonify({"error": "No email provided"}), 400
```
Same

DavidMStraub commented on gramps_webapi/auth/oidc.py:
```diff
@@ -0,0 +1,155 @@
+
+
+    if not user:
+        tree_id = get_tree_id("default")  # You can replace this logic as needed
```
That doesn't make sense, and anyway you're not adding that to the database.

DavidMStraub commented on gramps_webapi/auth/oidc.py:
```diff
@@ -0,0 +1,155 @@
+
+
+        try:
+            user = User(
```
There is an `add_user` function for that.

yuriBean commented:
User is added via add_user function now. replaced comments. replaced jsonify with abort_with_message.

adepssimius commented on gramps_webapi/auth/oidc.py:
```diff
@@ -0,0 +1,56 @@
+def configure_oauth(app):
+    oauth.init_app(app)
+
+    oauth.register(
```
Note that OIDC is just a defined way to use oauth. From the docs section that you linked:
> An OpenID Connect client is no different than a normal OAuth 2.0 client. When registered with openid scope, the built-in Flask OAuth client will handle everything automatically

DavidMStraub commented on gramps_webapi/auth/oidc.py:
```diff
@@ -0,0 +1,156 @@
+
+
+@oidc_bp.route("/login/<provider>")
+def login(provider):
```
Doc string please

DavidMStraub replied:
And please add an example to the doc string, I see there is a `redirect_uri` argument expected in the request? What should it be?

DavidMStraub commented on gramps_webapi/auth/oidc.py:
```diff
@@ -0,0 +1,156 @@
+        abort_with_message(403, "OAuth is not enabled")
+
+
+    redirect_uri = request.args.get("redirect_uri", url_for("oidc.authorize", provider=provider, _external=True))
```
Is `oidc.authorize` correct? Why not `gramps_webapi.auth.oidc.authorize`?

DavidMStraub commented on gramps_webapi/app.py:
```diff
@@ -122,31 +115,39 @@ def create_app(config: Optional[Dict[str, Any]] = None, config_from_env: bool =
     request_cache.init_app(app, config=app.config["REQUEST_CACHE_CONFIG"])
+
+
+    configure_oauth(app)
```
This will register the providers.

Does this involve an HTTP request to the provider endpoint? If so, what if that request fails? Will the whole flask app fail to start?

DavidMStraub commented:
I've added a couple more comments.

DavidMStraub commented:
To be clear, anyone helping bringing this to the finish line would be very welcome, I won't have time to directly work on this feature, but the current PR is far from being in a mergeable state.

DavidMStraub commented:
Closing in favor of #691.
